### PR TITLE
feat: adds support for entity key generator

### PR DESCRIPTION
**Pull request by octonato**
References https://github.com/lightbend/kalix/issues/7790

### CODE REVIEW

File: samples/spring-valueentity-counter/src/main/java/com/example/CounterEntity.java [hunk 1]
```diff
@@ -8,12 +8,12 @@
...
 @EntityType("counter")                                     // <1>
-@EntityKey("counter_id")                                   // <2>
-public class CounterEntity extends ValueEntity<Integer> {  // <3>
+public class CounterEntity extends ValueEntity<Integer> {  // <2>
 
   @Override
-  public Integer emptyState() { return 0; }                  // <4>
+  public Integer emptyState() { return 0; }                  // <3>
 
+  @EntityKey("counter_id")                                   // <4>
```
Comment (octonato) [inline samples/spring-valueentity-counter/src/main/java/com/example/CounterEntity.java:16]: Now that EntityKey can be on a method, it makes it easier to put it next to the PostMapping. At least for the getting started guide where we don't want to be too verbose.

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala [hunk 2]
```diff
@@ -16,23 +16,55 @@
...
-        KalixMethod(restMethod, entityKeys = entityKeys).withKalixOptions(buildJWTOptions(restMethod.javaMethod))
+
+        val entityKeyOnMethod = restMethod.javaMethod.getAnnotation(classOf[EntityKey])
+        val generateEntityKey = restMethod.javaMethod.getAnnotation(classOf[GenerateEntityKey])
+
+        if (entityKeyOnMethod != null && generateEntityKey != null)
+          throw ServiceIntrospectionException(
+            restMethod.javaMethod,
+            "Invalid annotation usage. Found both @EntityKey and @GenerateEntityKey annotations. " +
+            "A method can only be annotated with of them, but not both.")
```
Comment (octonato) [inline sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:51]:
```suggestion
            "A method can only be annotated with one of them, but not both.")
```

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+ * A method annotated with this annotations should not be annotated with {@link EntityKey},
+ * if it does, an error will be raised.
+ *
+ * The generated key will be a Version 4 (random) UUID. The UUID will be generated using a cryptographically secure
+ * random number generator.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface GenerateEntityKey {
```
Comment (octonato) [inline sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:34]: The documentation for this quite hidden in the gRPC SDK, I think we should add it on the entities pages. 

I'm planning to do it on follow up PR and then also cover Value Entities.

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java [hunk 4]
```diff
@@ -75,6 +75,46 @@ private Integer privateMethodSimilarSignature(Integer event) {
...
+    @PutMapping("/eventsourced/int/{number}")
+    public Integer getInteger(@PathVariable Integer number) {
+      return number;
+    }
+  }
+
+  @EntityType("counter")
+  public static class IllDefinedEntityWithEntityKeyGeneratorAndEntityKey extends EventSourcedEntity<Integer> {
+    @GenerateEntityKey
+    @EntityKey("id")
```
Comment (octonato) [inline sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:111]: this should be illegal, method can't have both

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java [hunk 5]
```diff
@@ -75,6 +75,46 @@ private Integer privateMethodSimilarSignature(Integer event) {
...
+    @GetMapping("/eventsourced/{counter_id}/int/{number}")
+    public Integer getInteger(@PathVariable Integer number) {
+      return number;
+    }
+  }
+
+  @EntityType("counter")
+  public static class CounterEventSourcedEntityWithEntityKeyGenerator extends EventSourcedEntity<Integer> {
+    @GenerateEntityKey
+    @PutMapping("/eventsourced/int/{number}")
```
Comment (octonato) [inline sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:102]: No EntityKey! Get it generated.

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java [hunk 6]
```diff
@@ -75,6 +75,46 @@ private Integer privateMethodSimilarSignature(Integer event) {
...
+    public Integer getInteger(@PathVariable Integer number) {
+      return number;
+    }
+  }
+
+  @EntityKey("id")
+  @EntityType("counter")
+  public static class CounterEventSourcedEntityWithEntityKeyMethodOverride extends EventSourcedEntity<Integer> {
+
+    @EntityKey("counter_id")
```
Comment (octonato) [inline sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:92]: Question is, what should we do when we have EntityKey on type level and on method level? 

I think a sensible default it to let pick the method one and override the type one. But open for other suggestions.

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+  Type value() default Type.VERSION_4_UUID;
+
+  enum Type {
+    /**
+     * Generate a Version 4 (random) UUID. The UUID will be generated using a cryptographically secure random
+     * number generator.
+     */
+    VERSION_4_UUID
```
Comment (octonato) [inline sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:43]: In gRPC, we need to have a value. In Java, we could just have an annotation. 

Not sure if we should keep this. Looking for ideas / feedback.

File: samples/spring-valueentity-counter/src/main/java/com/example/CounterEntity.java (hunk 1 above)
Comment (franciscolopezsancho) [reply samples/spring-valueentity-counter/src/main/java/com/example/CounterEntity.java:16]: I like this. It separates, even more, the idea of 'service' from the 'entity'.

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala [hunk 8]
```diff
@@ -16,23 +16,55 @@
...
-    val entityKeys = component.getAnnotation(classOf[EntityKey]).value()
+
+    val entityKeysOnType = {
+      val anno = component.getAnnotation(classOf[EntityKey])
+      if (anno != null) anno.value()
+      else Array.empty[String]
+    }
 
     val kalixMethods =
       RestServiceIntrospector.inspectService(component).methods.map { restMethod =>
```
Comment (franciscolopezsancho) [inline sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:42]: If the user adds `EntityKey` or `GeneratedEntityKey` to more than one method the last method wins?
If this is the case, Would be better to add validation so only one method can have one annotation or the other?

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java (hunk 6 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:92]: Another possibility I can think of would be raising an error, not allowing two EntityKey annotations.

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala (hunk 8 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:42]: oh sorry. I misunderstood. Never mind. There will be one annotation on the type or one per method with GET/POST. Although now this makes me think. Would make sense to add validation that there is at least one instance of 'EntityKey' or 'GeneratedEntityKey' that the rest method can use?

File: sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java (hunk 6 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/test/java/kalix/springsdk/testmodels/eventsourcedentity/EventSourcedEntitiesTestModels.java:92]: It makes sense to me, yup.

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.lang.annotation.*;
+
+/**
+ * Indicates that Kalix should generate an entity key when this method is invoked.
+
+ * The generated entity key will be accessible via the entity context supplied in the SDK.
+ * A method annotated with this annotations should not be annotated with {@link EntityKey},
+ * if it does, an error will be raised.
+ *
```
Comment (franciscolopezsancho) [inline sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:27]:
```suggestion
 * The generated entity key is accessible via the entity context supplied in the SDK.
 * A method annotated with this annotation should not be annotated with {@link EntityKey},
 * if it does, an error will be raised.
 *
```

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala [hunk 10]
```diff
@@ -16,23 +16,55 @@
...
+          if (generateEntityKey != null) {
+            val entityOptions = kalix.EntityMethodOptions.newBuilder().setKeyGenerator(Generator.VERSION_4_UUID)
+            val methodOpts = kalix.MethodOptions.newBuilder().setEntity(entityOptions)
+            KalixMethod(restMethod).withKalixOptions(methodOpts.build())
+
+          } else {
+            // keys defined on Method level get precedence
+            val entityKeysToUse =
+              if (entityKeyOnMethod != null) entityKeyOnMethod.value()
+              else entityKeysOnType
```
Comment (octonato) [inline sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:63]: @franciscolopezsancho, concerning you previous comment. Here we can indeed validate it once more. If `entityKeysToUse` end up being empty that's an invalid method. 

I don't know if that's already validated down the road, but in any case, better to have the validation here. 

I will add it.

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java (hunk 3 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:34]: would you like to add the reference to an issue before closing this PR?

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala [hunk 11]
```diff
@@ -16,23 +16,63 @@
...
+
+            if (entityKeysToUse.isEmpty)
+              throw ServiceIntrospectionException(
+                restMethod.javaMethod,
+                "Invalid command method. No @EntityKey nor @GenerateEntityKey annotations found. " +
+                "A command method should be annotated with either @EntityKey or @GenerateEntityKey, or " +
+                "the an @EntityKey annotation should be present at class level.")
```
Comment (franciscolopezsancho) [inline sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:70]:
```suggestion
                "an @EntityKey annotation should be present at class level.")
```

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java (hunk 7 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:43]: I'm not sure I understand the problem.
`In gRPC, we need to have a value. In Java, we could just have an annotation.` I can't see any other elegant way to provide the users the ability to express that they want an auto-generated key. Another option would be adding a boolean property in the `Entity`? Such as `generateKey `?  But this doesn't match the current model of annotations in the SDK. Maybe if we were creating things with a [builder pattern](https://www.digitalocean.com/community/tutorials/builder-design-pattern-in-java) like in the java-SDK that would make sense. 
So all of that is to say that I would keep this implementation.

On top of that from what I understand this class `value` is never used and that is why you set it to a fixed value and with an enum with only one option. To inform the user this is it. There are no other options here. This makes sense to me. I quite like it.

**Review by franciscolopezsancho — APPROVED** [review]
LGTM

File: sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala (hunk 10 above)
Comment (franciscolopezsancho) [reply sdk/spring-sdk/src/main/scala/kalix/springsdk/impl/EntityDescriptorFactory.scala:63]: I can see you did

File: sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java (hunk 7 above)
Comment (octonato) [reply sdk/spring-sdk/src/main/java/kalix/springsdk/annotations/GenerateEntityKey.java:43]: A gRPC annotation can't stands on its own. It needs to have an assigned value. Even though we only have one type, UUID V4, we need to write `option (kalix.method).entity.key_generator = VERSION_4_UUID`. 

The Java annotation could have been just `@GenerateEntityKey` without any value.
Basically, the enum I introduced is not needed and is there more to align with the gRPC counterpart. That's all.

**Comment by efgpinto** [comment]
@octonato did you consider having a property of `@EntityKey` instead of a new annotation (as mentioned by Cisco above)? I feel like this concept is so close to an EntityKey, that I was wondering if we really should introduce a new annotation for it.. Just concerned that we start having too fine grained Annotations.

**Pull request merged**